### PR TITLE
[RAD-1034] Transfer legs

### PR DESCRIPTION
**Pull request by michaz**
Insert TRANSFER legs, analogous to current ACCESS/EGRESS legs, between any two pt legs directly following one another where the second one starts elsewhere than where the first one ends.

Implement by just routing on the fly -- I currently strongly suspect that this doesn't matter that much, performance-wise, so for now, I'm not pre-computing them and saving them somewhere.
(Simple CH routing is so inexpensive that for short routes it can basically be considered a look-up -- if I'm wrong let's improve later).

On my way, I changed the logic for the access/egress legs a bit -- I *think* equivalently -- so that we don't expect walk legs returned by GH specifically and only at the first and last position. In this PR, we now convert any walk leg that we may get, and only mark them as ACCESS/EGRESS on first/last position. This is to prepare for GH eventually returning specific transfer legs directly.

I tried to set up my testing infrastructure but ran into an issue -- I think our deserializer refuses to deserialize our current modified response schema (or something), so I left the test commented out -- still putting it in so I can use it locally to trigger things. ,-)

### CODE REVIEW

**Comment by michaz** [comment]
In the process I removed the bit where we were removing edge ids of the last leg from the first leg, because I think we found the root cause of this in the meantime (?) and I wasn't sure how to translate this to the there-can-be-more-than-two-walk-legs situation.... Please verify.

**Comment by kaelgreco** [comment]
@mark-idleman how much work would this be to update and include in the mew protos? It would simplify the client side a lot!

**Comment by mark-idleman** [comment]
> @mark-idleman how much work would this be to update and include in the mew protos? It would simplify the client side a lot!

Errrr....this is a bit of a can of worms. There will need to be a decent amount of testing with this big of a change, and we'd need to change the GRPC code to include this new logic as well.

I'd strongly prefer getting in the big GRPC java PR and the first proto PR before digging into this. We're getting to the point where we need to start merging stuff so that I can add in these features in a clean way, continuing to pile on to the mega-grpc PR probably isn't a good call anymore haha.

**Comment by mark-idleman** [comment]
Following up on my last reply, I started to look over this more today and would be happy to get it updated based on recent changes/tested/merged as my next task; I'd still probably like it if we could get the big GRPC PR merged first though, just to make merging this change easier. I can get started by cleaning up/testing the HTTP version that's implemented here, though (the updates to the GRPC codebase should basically be analogous)

**Comment by mark-idleman** [comment]
@michaz , I did some clean-up and testing here, and overall I think the transfer leg logic is working well! The only issue I've noticed is that we're once again seeing the issue where the last foot leg (EGRESS) always contains all of the stable IDs from the first foot leg (ACCESS) - you mentioned that this might have been fixed so you removed the logic that [we had before](https://github.com/replicahq/graphhopper/pull/40/files#diff-ee52d62b9c7e99b89de34d3f601fdb337567f9fd8999124860089d09c36047d0L114-L115) for deduping this, but it looks like it will need to be added back, which I'll do now.

File: grpc/src/main/java/com/replica/RouterImpl.java [hunk 1]
```diff
@@ -295,38 +296,65 @@ public void routePt(PtRouteRequest request, StreamObserver<PtRouteReply> respons
...
-
-                Trip.WalkLeg firstLeg = (Trip.WalkLeg) walkLegs.get(0);
-                Trip.WalkLeg lastLeg = (Trip.WalkLeg) walkLegs.get(1);
-
-                List<String> lastLegStableIds = lastLeg.details.get("stable_edge_ids").stream()
-                        .map(idPathDetail -> (String) idPathDetail.getValue())
-                        .filter(id -> id.length() == 20)
-                        .collect(toList());
-
-                // The first leg contains stable IDs for both walking legs for some reason,
```
Comment (mark-idleman) [inline grpc/src/main/java/com/replica/RouterImpl.java:317]: One note here - this hacky code was removed in this PR, and I was a bit worried about it, but I checked and verified that it no longer seems to be needed. It seems to be the case that the bug we were addressing here no longer is a problem in graphhopper (I added a unit test just to be sure that all stable edge IDs across all legs are unique as well).

@michaz , does this sound right to you?

File: grpc/src/main/java/com/replica/RouterImpl.java [hunk 2]
```diff
@@ -295,38 +296,65 @@ public void routePt(PtRouteRequest request, StreamObserver<PtRouteReply> respons
...
+                                            Date.from(thisLeg.getArrivalTime().toInstant().plusMillis(transferPath.getTime()))
+                                    );
+                                    path.getLegs().add(new CustomWalkLeg(transferLeg, fetchWalkLegStableIds(transferLeg), "TRANSFER"));
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // ACCESS legs contains stable IDs for both ACCESS and EGRESS legs for some reason,
```
Comment (kaelgreco) [inline grpc/src/main/java/com/replica/RouterImpl.java:349]: ?
as in, it's the reverse direction stable edge ids, or the other leg's stable ids?

**Review by kaelgreco — APPROVED** [review]
sweeeet this is looking good to me! (@michaz if you have a chance to give this a pass too it'd be greatly appreciated)

let's talk offline about how to stage all of these changes (gh/idls/replica client) so things go smoothly

File: analyze_functional_test_results.py [hunk 3]
```diff
@@ -61,13 +61,14 @@ def calculate_query_time_90th_percentile(response_set: dict) -> float:
 
 def calculate_transit_ratio(response: dict) -> float:
     first_path = response["paths"][0]
+    transit_legs = filter(lambda x: x.has_key("transit_metadata"), first_path["legs"])
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply analyze_functional_test_results.py:64]: Ah shoot I always forget this is possible...will update, thanks!

**Pull request merged**